lib/db_connection.ex

```diff
@@ -953,243 +987,235 @@ defmodule DBConnection do
-    handle(conn, status, conn_state, :handle_close, [query], opts)
+  defp execute(conn, query, params, meter, opts) do
+    run_execute(conn, get_info(conn), query, params, meter, opts)
   end
 
-  defmacrop time() do
-    if function_exported?(:erlang, :monotonic_time, 0) do
-      quote do: :erlang.monotonic_time()
-    else
-      quote do: :os.timestamp()
-    end
+  defp run_execute(conn, conn_info, query, params, meter, opts) do
+    meter = event(meter, :execute)
+    handle(conn, conn_info, :handle_execute, [query, params], meter, opts)
   end
 
-  defp run_meter(%DBConnection{} = conn, fun, opts) do
-    case Keyword.get(opts, :log) do
-      nil ->
-        {run(conn, fun, opts), nil}
-      log ->
-        run_meter(conn, log, [], fun, opts)
-      end
-  end
-  defp run_meter(pool, fun, opts) do
-    case Keyword.get(opts, :log) do
-      nil ->
-        {run(pool, fun, opts), nil}
-      log ->
-        run_meter(pool, log, [checkout: time()], fun, opts)
+  defp raised_close(conn, query, meter, opts, kind, reason, stack) do
+    conn_info = get_info(conn)
+    with {:ok, _, meter} <- run_close(conn, conn_info, query, meter, opts) do
+      {kind, reason, stack, meter}
     end
   end
 
-  defp run_meter(conn, log, times, fun, opts) do
-    fun = fn(conn2) ->
-      start = time()
-      result = fun.(conn2)
-      stop = time()
-      meter = {log, [stop: stop, start: start] ++ times}
-      {result, meter}
-    end
-    run(conn, fun, opts)
+  defp run_close(conn, conn_info, query, meter, opts) do
+    meter = event(meter, :close)
+    handle_cleanup(conn, conn_info, :handle_close, [query], meter, opts)
   end
 
-  defp decode_log(_, _, _, nil, result), do: log_result(result)
-  defp decode_log(call, query, params, {log, times}, result) do
-   log(call, query, params, log, [decode: time()] ++ times, result)
+  defp handle_cleanup(_conn, :closed, _callback, _args, meter, _opts) do
+    err = DBConnection.ConnectionError.exception("connection is closed")
+    {:error, err, meter}
   end
-
-  defp transaction_log(nil), do: :ok
-  defp transaction_log({log, times, callback, result}) do
-    call = transaction_call(callback)
-    result = transaction_result(result)
-    _ = log(:transaction, call, nil, log, times, result)
-    :ok
+  defp handle_cleanup(conn, conn_info, callback, args, meter, opts) do
+    {status, _} = conn_info
+    handle(conn, conn_info, callback, args, meter, opts, status)
   end
 
-  defp transaction_call(:handle_begin), do: :begin
-  defp transaction_call(:handle_commit), do: :commit
-  defp transaction_call(:handle_rollback), do: :rollback
+  defp run(%DBConnection{} = conn, fun, arg, meter, opts) do
+    fun.(conn, get_info(conn), arg, meter, opts)
+  end
+  defp run(pool, fun, arg, meter, opts) do
+    meter = event(meter, :checkout)
+    with {:ok, conn, conn_info, meter} <- checkout(pool, meter, opts) do
+      try do
+        fun.(conn, conn_info, arg, meter, opts)
```

Comment:
All "connection time" runs inside this fun (or similar, above/below), which could be instrumented using fun based instrumentation.